src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php

```diff
@@ -121,6 +115,18 @@ public function collect(Request $request, Response $response, \Exception $except
             $this->data['request_request']['_password'] = '******';
         }
 
+        foreach ($this->data as $key => $value) {
+            if (!is_array($value)) {
+                continue;
+            }
+            if ('request_headers' === $key || 'response_headers' === $key) {
+                $value = array_map(function ($v) { return isset($v[0]) && !isset($v[1]) ? $v[0] : $v; }, $value);
+            }
+            if ('request_server' !== $key && 'request_cookies' !== $key) {
+                $this->data[$key] = $value;
+            }
+        }
+
```

Comment:
Above should apply to 3.1 as well.

In master this is already applied due #20595.